### PR TITLE
[IMP] web_editor: remove assets from documents

### PR DESCRIPTION
**Pull request by fja-odoo**
"Document" tab in "Select a media" Dialog was displaying all the assets
files. Now all files containing "assets_" are excluded from the
"Document" tab. This is similar to the regenerateAssets function in
debug_manager that deletes assets based on this criteria.

For consistency the Assets Domain is now a utils.js function

task-1945874
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ var Dialog = require('web.Dialog');
 var fonts = require('wysiwyg.fonts');
 var Widget = require('web.Widget');
 var concurrency = require('web.concurrency');
+var session = require('web.session');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js]: Move this in another PR targeting the first version in which the bug is present.

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js [hunk 2]
```diff
@@ -276,6 +277,9 @@ var ImageWidget = MediaWidget.extend({
         } else {
             domain = domain.concat(attachedDocumentDomain);
         }
+        if (this.options.document) {
+            domain.push('&', ['name', 'not like', 'assets_']);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js]: I don't see how this could work... if `domain` is `['|', A, B]`, after this line it will become `['|', A, B, '&', C]`, which is not a valid domain... is this working ? Try in cases where there are documents in the database

Anyways, try to share the domain targeting assets (where you found it for the regenerate button) and use it here with a correct combination of domains :wink: Maybe in the utils.js / misc.js file in web if no better option.

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js (hunk 1 above)
Comment (fja-odoo) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js]: here is the new PR: https://github.com/odoo/odoo/pull/33247  can you assign someone to it ?

File: addons/web/static/src/js/core/utils.js [hunk 3]
```diff
@@ -737,6 +737,17 @@ var utils = {
...
+     * return the domain targeting assets files
+     *
+     * @returns {Array} Domain of assets files
+     */
+    assetsDomain: function () {
+        return [
+            ['res_model', '=', 'ir.ui.view'],
+            ['name', 'like', 'assets_']
+        ];
+    }
```
Comment (qsm-odoo) [inline addons/web/static/src/js/core/utils.js]: I am being strict but:
- "Returns the domain targeting assets files." (Uppercase, ending point, conjugated verb)
- Trailing comma
- I know it was there before, but as you touch the code around: there is an useless blank line after it remove it :)

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js [hunk 4]
```diff
@@ -277,14 +278,17 @@ var ImageWidget = MediaWidget.extend({
...
         domain.push('|',
             ['mimetype', '=', false],
             ['mimetype', this.options.document ? 'not in' : 'in', ['image/gif', 'image/jpe', 'image/jpeg', 'image/jpg', 'image/gif', 'image/png']]);
         if (needle && needle.length) {
             domain.push('|', ['datas_fname', 'ilike', needle], ['name', 'ilike', needle]);
         }
         domain.push('|', ['datas_fname', '=', false], '!', ['datas_fname', '=like', '%.crop'], '!', ['name', '=like', '%.crop']);
+        // the assets should not be part of the documents
+        if (this.options.document) {
+            domain = domain.concat('!', '&', utils.assetsDomain());
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js]: Creating domains in JS is bad I know :) Let's put the `&` inside the `assetsDomain` implementation in utils.js so that this domain can more easily be combined with other domains.

File: addons/web/static/src/js/core/utils.js (hunk 3 above)
Comment (qsm-odoo) [reply addons/web/static/src/js/core/utils.js]: You forgot about the trailing comma.

Comment (fja-odoo) [reply addons/web/static/src/js/core/utils.js]: I think you are looking at the wrong version I did add a comma there.

Comment (qsm-odoo) [reply addons/web/static/src/js/core/utils.js]: `},` you added the comma in the array which I don't think we do actually

File: addons/web/static/src/js/core/utils.js [hunk 5]
```diff
@@ -737,7 +737,18 @@ var utils = {
...
+    /**
+     * Returns the domain targeting assets files.
+     *
+     * @returns {Array} Domain of assets files
+     */
+    assetsDomain: function () {
+        return [
+            '&',
+            ['res_model', '=', 'ir.ui.view'],
+            ['name', 'like', 'assets_']
```
Comment (JKE-be) [inline addons/web/static/src/js/core/utils.js]: =like

File: addons/web/static/src/scss/utils.scss [hunk 6]
```diff
@@ -339,3 +339,41 @@
...
+    background-size: cover;
+}
+
+%o-nocontent-empty-document {
+    @extend %o-nocontent-init-image;
+    @include size(120px, 80px);
+    margin-top: 30px;
+    margin-bottom: 30px;
+    background: transparent url(/web/static/src/img/empty_folder.svg) no-repeat center;
+}
```
Comment (qsm-odoo) [inline addons/web/static/src/scss/utils.scss]: Every normal editor should automatically add a end-of-file line feed, could you check your options ?

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js [hunk 7]
```diff
@@ -224,7 +225,11 @@ var ImageWidget = MediaWidget.extend({
     _adaptLoadMore: function () {
         var noMoreImgToLoad = this.IMAGES_DISPLAYED_TOTAL >= this.records.length;
         this.$('.o_load_more').toggleClass('d-none', noMoreImgToLoad);
-        this.$('.o_load_done_msg').toggleClass('d-none', !noMoreImgToLoad);
+        if (this.records.length > 0) {
+            this.$('.o_load_done_msg').toggleClass('d-none', !noMoreImgToLoad);
+        } else {
+            this.$('.o_load_done_msg').toggleClass('d-none', true);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js]: `this.$('.o_load_done_msg').toggleClass('d-none', !this.records.length || !noMoreImgToLoad)` ?

File: addons/web_editor/static/src/xml/wysiwyg.xml [hunk 8]
```diff
@@ -185,6 +185,14 @@
     </t>
     <t t-name="wysiwyg.widgets.files.existing.content">
         <div class="existing-attachments">
+            <div t-if="rows.length === 0" class="o_nocontent_help">
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/wysiwyg.xml]: `!rows.length` is our convention

File: addons/web_editor/static/src/xml/wysiwyg.xml [hunk 9]
```diff
@@ -185,6 +185,14 @@
...
+                <p class="o_empty_folder_image">
+                    <t t-if="isDocument">No documents found</t>
+                    <t t-else="">No images found</t>
+                </p><p>
+                    You can upload <t t-if="isDocument">documents</t><t t-else="">images</t> with the button located in the top left of the screen
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/wysiwyg.xml]: This is not translatable. Use string formatting or duplicate the string for both type of files.

File: addons/web_editor/static/src/xml/wysiwyg.xml [hunk 10]
```diff
@@ -185,6 +185,19 @@
     </t>
     <t t-name="wysiwyg.widgets.files.existing.content">
         <div class="existing-attachments">
+            <div t-if="!rows.length" class="o_nocontent_help">
+                <p class="o_empty_folder_image">
+                    <t t-if="isDocument">No documents found</t>
+                    <t t-else="">No images found</t>
+                </p><p>
+                <t t-if="isDocument">
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/xml/wysiwyg.xml]: Indentation problem here... change that and we can merge

**Comment by fja-odoo** [comment]
@qsm-odoo True I fixed it !

File: addons/web_editor/static/src/js/wysiwyg/widgets/media.js [hunk 11]
```diff
@@ -426,8 +426,9 @@ var FileWidget = SearchableMediaWidget.extend({
 
         // adapt load more
         var noMoreImgToLoad = this.numberOfAttachmentsToDisplay >= this.attachments.length;
-        this.$('.o_load_more').toggleClass('d-none', noMoreImgToLoad);
-        this.$('.o_load_done_msg').toggleClass('d-none', !noMoreImgToLoad);
+        var $loadMoreButton = this.$('.o_load_more');
+        this.$('.o_load_done_msg').toggleClass('d-none', $loadMoreButton.hasClass('d-none') || !noMoreImgToLoad);
+        $loadMoreButton.toggleClass('d-none', noMoreImgToLoad);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/media.js]: @fja-odoo It my not written like that before I reviewed right ? I find it really difficult to understand what this does. The new state depends on the old state + some condition... while it should only depend on the number of total images and the number of displayed images.

Comment (fja-odoo) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js]: Previously it was only depending on the  number of total images and the number of displayed images but the new specs ask to only display the message if the "Load more" button was previously displayed.

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/wysiwyg/widgets/media.js]: Can't the condition be replaced by `nbDisplayed <= minNbDisplayed` ?

**Comment by qsm-odoo** [comment]
@fja-odoo Please review and test your branch before sending it to review. The branch is not working, once I search something it tells me there is no result, then I can never search again.

**Comment by fja-odoo** [comment]
@qsm-odoo There are fixes for the image modal here that still need to be merged https://github.com/odoo/odoo/pull/33983
This commit in particular: b74132eb2ad028ae677117a828625414560e5022
Should I cherry pick it in this PR ?

**Comment by qsm-odoo** [comment]
Ok no we will have to wait for forward-port. In those cases, try blocking the task on odoo.com or leave it "gray". I'll check the other PR, thanks.

**Comment by fja-odoo** [comment]
@qsm-odoo I think this can be r+ now.
